### PR TITLE
Feature/user conversations

### PR DESCRIPTION
**Pull request by MatiasCab**
#### Board:
* [Ticket #NUMBER_OF_THE_TICKET](https://trello.com/c/KbzhUrEj)
---
#### Description:
A new conversation is created everytime the user matches with a new target.

Only one conversation is created between two users, even if they match across multiple targets.
---
#### Notes:
*Considerations and Assumptions:
• A match occurs only if both targets are within their respective radii.
• A target can only match with another target (only one match is allowed).
• Two users can only match with one target.
---
#### Tasks:
- [x] Add each element in this format
---
#### Risk:
*
---

### CODE REVIEW

File: app/models/conversation.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  private
+
+  def conversation_users
+    @conversation_users ||= users
+  end
+
+  def conversation_targets
+    @conversation_targets ||= targets
+  end
```
Comment (sebastiancaraballo) [inline app/models/conversation.rb]: I recommend extracting this logic to a service object. Models should be as thin as possible. cc @smerlo https://www.youtube.com/watch?v=CRboMkFdZfg&ab_channel=RubyCentral

File: app/services/create_target_service.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  def targets
+    @targets ||= @current_user.targets
+  end
+
+  def check_matchs(new_target)
+    matched_target = find_target_to_match(new_target)
+    return if matched_target.nil?
+
+    matched_target.update!(matched: true)
+    new_target.update!(matched: true)
```
Comment (sebastiancaraballo) [inline app/services/create_target_service.rb:26]: Just asking here, what does it mean that a target is `matched`? Is it able to match again?

File: spec/factories/match_users_conversations.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#
+#  index_match_users_conversations_on_conversation_id  (conversation_id)
+#  index_match_users_conversations_on_target_id        (target_id)
+#  index_match_users_conversations_on_user_id          (user_id)
+#
+FactoryBot.define do
+  factory :match_users_conversation do
+    association :user, factory: :user
+    association :conversation, factory: :conversation
+    association :target, factory: :target
```
Comment (sebastiancaraballo) [inline spec/factories/match_users_conversations.rb]:
```suggestion
    target
```

**Review by sebastiancaraballo — COMMENTED** [review]
Nice work. I recommend doing smaller PRs since this one implements multiple features.

**Review by smerlo — APPROVED** [review]
LGTM!

File: app/controllers/api/v1/targets_controller.rb [hunk 4]
```diff
@@ -7,12 +7,15 @@ def index
 
       def create
         authorize Target
-        @target = current_user.targets.create!(resource_params)
+        @target = CreateTargetService.new(current_user, resource_params).create!
+        return unless @target.matched
```
Comment (enzofab91) [inline app/controllers/api/v1/targets_controller.rb]:
```suggestion
        return unless @target.matched?
```

File: app/models/target.rb [hunk 5]
```diff
@@ -21,13 +22,36 @@ class Target < ApplicationRecord
...
   validates :radius, presence: true, numericality: { greater_than: 0 }
   validates :lat, :lng, presence: true, numericality: true
-  validate  :max_target_amount_reached, unless: -> { user.nil? }
+  validate  :max_target_amount_reached, unless: -> { user.nil? }, on: :create
 
   MAX_TARGETS_AMOUNT = ENV.fetch('TARGET_CREATION_LIMIT', '3').to_i
 
+  scope :by_topic, ->(topic) { where(topic:) }
+  scope :not_same_user, ->(user) { where.not(user:) }
+  scope :inside_their_radius, lambda { |lng, lat|
```
Comment (enzofab91) [inline app/models/target.rb:37]: Wow that's interesting, this is postgresql syntax? 👏 

Another possibility would have been to use [geokit](https://github.com/geokit/geokit-rails) gem, that helps you on geolocation calculations

File: app/services/get_conversation_data_service.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+class GetConversationDataService
+  def initialize(target)
+    @target = target
+    @user = target.user
+    @conversation = target.conversation
+  end
+
+  def other_user
+    first_conversation_user = conversation_users.first
+    return first_conversation_user if first_conversation_user.id != @user.id
```
Comment (enzofab91) [inline app/services/get_conversation_data_service.rb]: You can add `attr_reader :user, @conversation` and you avoid using the variables.

See https://www.rubyguides.com/2018/11/attr_accessor/

**Review by enzofab91 — APPROVED** [review]
Awesome work! 🎩

**Pull request merged**